protovalidate/validator.py

```diff
@@ -84,8 +84,34 @@ def collect_violations(
             constraint.validate(ctx, message)
             if ctx.done:
                 break
+        for violation in ctx.violations.violations:
+            if violation.HasField("field"):
+                violation.field.elements.reverse()
+            if violation.HasField("rule"):
+                violation.rule.elements.reverse()
```

Comment:
I'm assuming we're reversing these because we have a list from root -> this nested field, and we want to render these as field -> root — is that right?

separately, is the structure of these fields covered by the conformance tests, or is that something we ought to separately add?

Reply:
> I'm assuming we're reversing these because we have a list from root -> this nested field, and we want to render these as field -> root — is that right?

Field paths are constructed in reverse, as we're returning violations up through the stack. Prior to this PR, the string-based field path was prepended to at each step. After this PR, lists are appended to instead, and then those lists get reversed at the end. Prepending to a list is O(n) since it must shift all of the elements, so instead we append then reverse later (reverse is theoretically very cheap since it just needs to swap a bunch of pointers around.)

If we had our choice of data structure we could use a deque here instead, but we don't.

> separately, is the structure of these fields covered by the conformance tests, or is that something we ought to separately add?

Absolutely! Painstakingly, the WIP conformance tests add field path and rule path tests to *all* existing violation cases. The entire field path and rule path are tested to ensure they match the conformance test cases. This PR passes right now.